### ISSUE TITLE
Interface support generic

### ISSUE THREAD

**Issue by jasonni** [LachlanMcKee/gsonpath#69]
hi @LachlanMcKee 
I write a generic interface, but the code gen error

```
@AutoGsonAdapter
public interface ApiResponse<T> {
  List<T> result();
}
```

```
public final class ApiResponse_GsonPathModel implements ApiResponse {
  private final List<T> result;
  ...
}
```

**Comment by LachlanMcKee** [comment]
Hi @jasonni 

Can you please explain what you are trying to achieve with the interface you have described above?

The annotation processor needs to know what the generic type is for any annotated model. I presume that you are trying to extend this interface, or implement it on a concrete model?

Ideally what you are attempting to do should be done as follows:

```
interface ApiResponse<T> {
  List<T> result();
}
```

```
@AutoGsonAdapter
interface UserApiResponse extends ApiResponse<User> {

}
```

However upon trying to do this, I found that the processor still has an issue with the generics. I will have to investigate this.

There **is** a way to solve this within version 1.8.0. You can directly extend the `List` interface and specify a `rootField`:

```
@AutoGsonAdapter(rootField = "result")
interface UserApiResponse extends List<User> {
}
```

That will then generate the following:

```
public final class UserApiResponseList_GsonPathModel extends GsonPathElementList<UserDetails> implements UserApiResponseList {
  private final List<UserDetails> internalList;

  public UserApiResponseList_GsonPathModel(List<UserDetails> internalList) {
    this.internalList = internalList;
  }

  @Override
  protected List<UserDetails> getList() {
    return internalList;
  }
}
```

```
public final class UserApiResponseList_GsonTypeAdapter extends TypeAdapter<UserApiResponseList> {
  private static final int MANDATORY_INDEX_INTERNALLIST = 0;

  private static final int MANDATORY_FIELDS_SIZE = 1;

  private final Gson mGson;

  public UserApiResponseList_GsonTypeAdapter(Gson gson) {
    this.mGson = gson;
  }

  @Override
  public UserApiResponseList read(JsonReader in) throws IOException {

    // Ensure the object is not null.
    if (!isValidValue(in)) {
      return null;
    }
    java.util.List<User> value_internalList = null;
    boolean[] mandatoryFieldsCheckList = new boolean[MANDATORY_FIELDS_SIZE];

    int jsonFieldCounter0 = 0;
    in.beginObject();

    while (in.hasNext()) {
      if (jsonFieldCounter0 == 1) {
        in.skipValue();
        continue;
      }

      switch (in.nextName()) {
        case "result":
          jsonFieldCounter0++;

          // Ensure the object is not null.
          if (!isValidValue(in)) {
            break;
          }

          java.util.List<User> value_internalList_safe = mGson.getAdapter(new com.google.gson.reflect.TypeToken<java.util.List<User>>(){}).read(in);
          if (value_internalList_safe != null) {
            value_internalList = value_internalList_safe;
            mandatoryFieldsCheckList[MANDATORY_INDEX_INTERNALLIST] = true;

          } else {
            throw new gsonpath.JsonFieldMissingException("Mandatory JSON element 'internalList' was null for class 'UserApiResponseList_GsonPathModel'");
          }
          break;

        default:
          in.skipValue();
          break;
      }
    }


    in.endObject();

    // Mandatory object validation
    for (int mandatoryFieldIndex = 0; mandatoryFieldIndex < MANDATORY_FIELDS_SIZE; mandatoryFieldIndex++) {

      // Check if a mandatory value is missing.
      if (!mandatoryFieldsCheckList[mandatoryFieldIndex]) {

        // Find the field name of the missing json value.
        String fieldName = null;
        switch (mandatoryFieldIndex) {
          case MANDATORY_INDEX_INTERNALLIST:
            fieldName = "internalList";
            break;

        }
        throw new gsonpath.JsonFieldMissingException("Mandatory JSON element '" + fieldName + "' was not found for class 'UserApiResponseList_GsonPathModel'");
      }
    }
    return new UserApiResponseList_GsonPathModel(
      value_internalList
    );
  }

  @Override
  public void write(JsonWriter out, UserApiResponseList value) throws IOException {
  }
}
```

Please let me know if this solves your issue! :)

**Comment by jasonni** [comment]
Hi @LachlanMcKee 
This one is awesome! 
```
@AutoGsonAdapter(rootField = "result")
interface UserApiResponse extends List<User> {
}
```
Thanks your help!

**Comment by LachlanMcKee** [comment]
Glad that you find it useful!

I'm going to leave this issue open while I investigate the generics issue you have helped me uncover!

**Comment by LachlanMcKee** [comment]
Hi @jasonni 

I have just released version 1.8.1 that will now properly accept generics. You will be able to create a model like the one I suggested above. e.g.

```
interface ApiResponse<T> {
  List<T> getResult();
}
```

```
@AutoGsonAdapter
interface UserApiResponse extends ApiResponse<User> {

}
```

Let me know if you have any further suggestions or issues!
